### PR TITLE
Connect twilio phone stream with fixie session

### PR DESCRIPTION
**Pull request by JerryChocoY**
1. Define dummy PhoneAudioSource and Link
2. Update Twilio steam example to connect with Fixie session with PhoneAudioSource and Link

### CODE REVIEW

File: examples/twilio_stream_example.py [hunk 1]
```diff
@@ -24,25 +31,60 @@ async def websocket_handler(request):
             # Using the event type you can determine what type of msg you are receiving
             if data["event"] == "connected":
                 logging.info(f"Received connected message={msg}")
+                # Warm up the voice session by connecting to the server.
+                await client.warmup()
             if data["event"] == "start":
                 logging.info(f"Received start message={msg}")
+                # Not just warming up...Start the voice session.
+                if not args.warmup_only:
```
Comment (juberti) [inline examples/twilio_stream_example.py:39]: This isn't important in this server, you can remove any warmup_only handling here.

File: examples/twilio_stream_example.py [hunk 2]
```diff
@@ -24,25 +31,60 @@ async def websocket_handler(request):
...
-                    )
-                media_count = media_count + 1
+                await source.write(chunk)
             if data["event"] == "closed":
                 logging.info("Received close message={msg}")
                 await ws.close()
 
     logging.info("Websocket connection closed")
+    # Wait for the voice session to end.
+    await done.wait()
```
Comment (juberti) [inline examples/twilio_stream_example.py:51]: Don't think you need to wait on `done` here?

File: examples/twilio_stream_example.py [hunk 3]
```diff
@@ -24,25 +31,60 @@ async def websocket_handler(request):
...
                 chunk = base64.b64decode(payload)
-                if not media_count % 100:
-                    logging.info(
-                        f"(Received {media_count}th media message={msg} with payload size={len(chunk)}"
-                    )
-                media_count = media_count + 1
+                await source.write(chunk)
             if data["event"] == "closed":
                 logging.info("Received close message={msg}")
                 await ws.close()
```
Comment (juberti) [inline examples/twilio_stream_example.py:47]: Probably you should stop the session here.

File: examples/twilio_stream_example.py [hunk 4]
```diff
@@ -24,25 +31,60 @@ async def websocket_handler(request):
...
+    await client.stop()
     return ws
 
 
 if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--agent", "-a", type=str, default="dr-donut", help="Agent ID to talk to"
+    )
+    parser.add_argument("--tts-voice", "-tv", type=str, help="TTS voice ID to use")
```
Comment (juberti) [inline examples/twilio_stream_example.py:61]: We switched to -V in the other example for this.

File: examples/twilio_stream_example.py [hunk 5]
```diff
@@ -24,25 +31,60 @@ async def websocket_handler(request):
...
+    parser.add_argument(
+        "--warmup-only", "-w", action="store_true", help="Only connect to the server"
+    )
+    args = parser.parse_args()
+
+    # Get the default microphone and audio output device.
+    source = audio_phone.PhoneAudioSource()
+    sink = audio_local.LocalAudioSink()
```
Comment (juberti) [inline examples/twilio_stream_example.py:69]: Do you want to replace this with PhoneAudioSink too?

File: fixie_sdk/voice/audio_phone.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
+import asyncio
+from typing import AsyncGenerator
+
+import g711
+import numpy
+
+from fixie_sdk.voice import audio_base
+
+
+class PhoneAudioSink(audio_base.AudioSink):
```
Comment (juberti) [inline fixie_sdk/voice/audio_phone.py:10]: You might want to pass the socket in here so you can write to it.

File: examples/twilio_stream_example.py [hunk 7]
```diff
@@ -24,25 +31,60 @@ async def websocket_handler(request):
...
     return ws
 
 
 if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--agent", "-a", type=str, default="dr-donut", help="Agent ID to talk to"
+    )
+    parser.add_argument("--tts-voice", "-tv", type=str, help="TTS voice ID to use")
+    parser.add_argument(
```
Comment (juberti) [inline examples/twilio_stream_example.py:62]: Remove this as noted above.

**Review by juberti — COMMENTED** [review]
This is on the right track. Left some minor comments.

File: examples/twilio_stream_example.py [hunk 8]
```diff
@@ -24,25 +31,60 @@ async def websocket_handler(request):
...
+    sink = audio_local.LocalAudioSink()
+
+    # Set up the voice session parameters.
+    params = VoiceSessionParams(
+        agent_id=args.agent,
+        tts_voice=args.tts_voice,
+    )
+
+    # Create the client for the voice session.
+    client = VoiceSession(source, sink, params)
```
Comment (juberti) [inline examples/twilio_stream_example.py:78]: Oops, forgot to mention that you'll need a different session object for every websocket connection. So this session creation should move to the socket handler.

File: fixie_sdk/voice/audio_base.py [hunk 9]
```diff
@@ -32,6 +32,9 @@ def enabled(self, enabled: bool):
     async def stream(self):
         pass
 
+    async def write(self, data: bytes):
```
[the remark this answers is not in the document]
Comment (JerryChocoY) [reply fixie_sdk/voice/audio_base.py]: done

File: examples/twilio_stream_example.py (hunk 1 above)
Comment (JerryChocoY) [reply examples/twilio_stream_example.py]: done

File: examples/twilio_stream_example.py (hunk 2 above)
Comment (JerryChocoY) [reply examples/twilio_stream_example.py]: nice catch. I had a typo: s/closed/stop

Comment (JerryChocoY) [reply examples/twilio_stream_example.py]: done

File: examples/twilio_stream_example.py (hunk 3 above)
Comment (JerryChocoY) [reply examples/twilio_stream_example.py:51]: nice catch. also s/closed/stope

File: examples/twilio_stream_example.py (hunk 4 above)
Comment (JerryChocoY) [reply examples/twilio_stream_example.py]: done. There are two voice_example.py files and we should remove one

File: examples/twilio_stream_example.py (hunk 5 above)
Comment (JerryChocoY) [reply examples/twilio_stream_example.py]: plan to do later.

File: fixie_sdk/voice/audio_phone.py (hunk 6 above)
Comment (JerryChocoY) [reply fixie_sdk/voice/audio_phone.py:11]: done. It is kind of ugly to pass websocket to the Sink

File: examples/twilio_stream_example.py (hunk 7 above)
Comment (JerryChocoY) [reply examples/twilio_stream_example.py:66]: done

File: examples/twilio_stream_example.py (hunk 8 above)
Comment (JerryChocoY) [reply examples/twilio_stream_example.py]: nice catch. I was not clear on py threading.  Moved all objs within websocket_handler, which is one per call

File: fixie_sdk/voice/audio_phone.py (hunk 6 above)
Comment (juberti) [reply fixie_sdk/voice/audio_phone.py:11]: I see what you are saying. I think this won't be a concern if you move this class to the example.py file, which is what I was thinking - that is, that this source and sink are specific to the Twilio scenario.

Open to other ideas though.

File: fixie_sdk/voice/audio_phone.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        self._ws = ws
+
+    async def start(self, sample_rate: int = 8000, num_channels: int = 1):
+        pass
+
+    async def write(self, chunk: bytes) -> None:
+        # TODO: write to ws
+        # https://github.com/twilio/media-streams/blob/master/node/connect-basic/server.js
+        # (Received 100th media message=WSMessage(type=<WSMsgType.TEXT: 1>, data='{"event":"media","sequenceNumber":"102","media":{"track":"inbound","chunk":"101","timestamp":"2088","payload":"/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////w=="},"streamSid":"MZ351c981a36cd9426fc1b17974b5d0327"}', extra='') with payload size=160
+        # frame_data = wav.read(1024)
```
Comment (juberti) [inline fixie_sdk/voice/audio_phone.py:25]: The chunk will just be a byte blob so you can just process it directly here, no need to read it using some other package.

File: examples/twilio_stream_example.py [hunk 11]
```diff
@@ -24,25 +36,42 @@ async def websocket_handler(request):
...
-                media_count = media_count + 1
-            if data["event"] == "closed":
-                logging.info("Received close message={msg}")
+                await source.write(chunk)
+            if data["event"] == "stop":
+                logging.info(f"Received stop message={msg}")
+                await client.stop()
                 await ws.close()
 
     logging.info("Websocket connection closed")
```
Comment (juberti) [inline examples/twilio_stream_example.py:53]: If the session is still alive, you should stop it here (ie if the socket gets interrupted somehow)

File: examples/twilio_stream_example.py (hunk 4 above)
Comment (juberti) [reply examples/twilio_stream_example.py]: oh, thanks for mentioning that. removed the old one

**Comment by juberti** [comment]
Looking good! If you want to land the phone sink in a separate PR, that's fine. Main remaining this for this PR would just be to move the phone source/sinks to the example file.

File: fixie_sdk/voice/audio_phone.py (hunk 10 above)
Comment (JerryChocoY) [reply fixie_sdk/voice/audio_phone.py]: ack

File: examples/twilio_stream_example.py (hunk 11 above)
Comment (JerryChocoY) [reply examples/twilio_stream_example.py:79]: done

**Comment by JerryChocoY** [comment]
> Looking good! If you want to land the phone sink in a separate PR, that's fine. Main remaining this for this PR would just be to move the phone source/sinks to the example file.

Added the sink. Will debug later: PhoneAudioSink.write is never called.

File: fixie_sdk/voice/audio_phone.py [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    def __init__(self, ws: aiohttp.web.WebSocketResponse) -> None:
+        super().__init__()
+        self._ws = ws
+        self._stream_sid = ""
+        self._sequence_no = 0
+
+    async def start(self, sample_rate: int = 8000, num_channels: int = 1):
+        pass
+
+    def stream_sid(self, value):
```
Comment (juberti) [inline fixie_sdk/voice/audio_phone.py:25]: make this a python `@property`

File: fixie_sdk/voice/audio_phone.py (hunk 6 above)
Comment (juberti) [reply fixie_sdk/voice/audio_phone.py:13]: Just move these classes to `twilio_stream_example.py` and this PR is good to go.

**Review by juberti — APPROVED** [review]
LG once remaining comments are resolved

File: fixie_sdk/voice/audio_phone.py (hunk 12 above)
Comment (JerryChocoY) [reply fixie_sdk/voice/audio_phone.py]: done

**Comment by JerryChocoY** [comment]
moved these classes to twilio_stream_example.py 

how to debug:
1. the worker receives the phone audio
2. the sink is called from the room

**Comment by JerryChocoY** [comment]
updated to audioop, which we can replace later, to make the source work. I can see the message now. The sink audio does not work yet
```
Agent: Merry Christmas, this is Santa Claus. Now tell me, have you been good this year?
User:  Good morning. Can you hear me?
Agent: Good morning! Yes, I can hear you loud and clear. What's your name?
User:  It's very slow.
Agent: I'm sorry to hear that. Sometimes the connection at the North Pole can be a bit tricky with all the snowstorms. 
```

**Comment by juberti** [comment]
Makes sense, we can proceed with that for now.

**Pull request merged**